### PR TITLE
Add support for more Dtypes

### PR DESCRIPTION
**Pull request by nickgg**
Add support for more integer types (char, byte, short, long) and floating point types (half, double) in the IR and implementations in the interpreter, llvm and cuda backends (inc. type propagation, etc).

A few details:
* `Int32` and `Float32` Dtypes are now just `Int` and `Float`.
* `Dtype::scalar_type()` now returns the `ScalarType` enum rather than a fixed Dtype object with lanes == 1, but I kept the old way for compatibility (esp. tests).
* This uses the type forall macro from c10 extensively to ensure our types are synchronized with the rest of PyTorch, but it will affect greppability a bit.
* `half` floats are auto converted to float in the generated cuda code (which is what the existing fuser does too), but I plan to go back and try to get real half math working on the GPU.

Test plan:
* Ran C++ tests, added a bunch of new ones.

### CODE REVIEW

File: test/cpp/tensorexpr/padded_buffer.h [hunk 1]
```diff
@@ -17,11 +17,42 @@ struct DefaultPaddedValue<int> {
...
+};
+
 template <>
 struct DefaultPaddedValue<float> {
   static constexpr float kValue = 0.1357;
 };
 
+template <>
+struct DefaultPaddedValue<at::Half> {
+  // at::Half ctor isnt constexpr, so just fill it with bits.
```
Comment (bertmaher) [inline test/cpp/tensorexpr/padded_buffer.h]: nit: `isn't`

File: torch/csrc/jit/tensorexpr/eval.h [hunk 2]
```diff
@@ -48,46 +51,54 @@ class Value {
...
 }
+AT_FORALL_SCALAR_TYPES_AND(Half, VALUE_AS_VEC_DISPATCH);
+#undef VALUE_AS_VEC_DISPATCH
 
 template <typename T>
 class PaddedBuffer;
 
-inline int mod_value(int lhs, int rhs) {
+template <typename T>
+inline typename std::enable_if<std::is_integral<T>::value, T>::type mod_value(
```
Comment (bertmaher) [inline torch/csrc/jit/tensorexpr/eval.h:85]: Any diff using `type_traits` is a good diff in my book <3

**Review by bertmaher — APPROVED** [review]
This is **so** cool.  Also I had no idea dtypes was going to be this involved, this diff is a real tour de force :-D.

Given it's such a big diff I think we should go ahead and get it in to minimize rebase pain.  I read through it quickly and things lgtm.

**Review by bertmaher — CHANGES_REQUESTED** [review]
Err, I just pulled this and it's segfaulting on my llvm-only box.

**Comment by nickgg** [comment]
@bertmaher thanks, should be fixed.

**Review by bertmaher — APPROVED** [review]
Works great, thanks!

**Pull request merged**